### PR TITLE
time-frequency analysis

### PR DESCRIPTION
**Pull request by pettetmw**
Here is the code for the latest group visualizations and stats, please note that we are working on dataset-tone branch, not master.
It took a while to remember how to make a jupyter notebook, so the tfr_pipe.ipynb in my github fork doesn't work. I've got a version that does, but still needs editing.  I'll push that before we meet.

### CODE REVIEW

**Review by ktavabi — CHANGES_REQUESTED** [review]
Considering this repo is not an application I'd like to see a few improvements to this commit before merging. 

- [x] Convert extensive comments to _useful_ notes chunk in the docstring at the top.
- [ ] Lose the local (MATLAB like) [functions](https://github.com/ktavabi/badbaby/pull/5/files#diff-49ef6ddbc0e3ab12d5e9e19d65d828d9R25) and just write out the few lines of code.
- [ ] commit only the files I or someone else needs to run on the data. 
- [ ] Archive junk [files](badbaby/python/fdrgist.py) and [extended commented sections](https://github.com/ktavabi/badbaby/pull/5/files#diff-49ef6ddbc0e3ab12d5e9e19d65d828d9R285)

File: badbaby/python/Makefile [hunk 1]
```diff
@@ -24,3 +41,7 @@ epofif/%b-epo.fif : tcircevo/%b-ave.fif tcircevo/%a-ave.fif ../tone/*/epochs/%b-
 tcircevo/%-ave.fif : ../tone/*/epochs/%-epo.fif
 	echo makeTcircevoFromEpo $@ $<
 
+# to automatically populate tcircevo when only prereqs exist
+newtcircevo : ../tone/*/epochs/*a-epo.fif ../tone/*/epochs/*b-epo.fif
+	ls $? | xargs basename -a | xargs printf %.20s\\n | xargs -I _FN_ make -s tcircevo/_FN_-ave.fif
+
```
Comment (ktavabi) [inline badbaby/python/Makefile:47]: Is this file necessary for scripts to run on dataset on another WS?

Comment (pettetmw) [reply badbaby/python/Makefile]: Nope, deleted in this commit

**Comment by pettetmw** [comment]
This notebook should be able to build everything from original tone data directory.  See the section on path and file names.

**Comment by ktavabi** [comment]
@pettetmw let's break this notebook up into a script per cell; starting with  

- [ ] write_subject_tfrs.py
- [ ] viz_grp_assr.py that also write TIDY csv

In the process, you can work on not repeating yourself in code, and I can move towards BIDS compliance in a modular fashion.

**Comment by pettetmw** [comment]
@ktavabi ,  Revising repetitive code should be straightforward enough.  About file saving loop, I have a couple of questions:

Data returned by TFR_Morlet are of type AverageTFR and EpochsTFR, which differ from the Evoked and Epochs types in the two key features:

1) They are saved as hd5.
2) They are not represented in MNEFUN file system tree; the closest match I could find in mnefun:write_epochs.py was the following:

    for si, subj in enumerate(subjects):
        epochs_dir = op.join(p.work_dir, subj, p.epochs_dir)
        evoked_dir = op.join(p.work_dir, subj, p.inverse_dir)

Since there is no established mnefun convention, I guess we can create a dedicated folder.  LMK what you prefer.

Also, please specify what's going to be in "picks" in the new loop? IDs, filenames, full paths?

**Comment by ktavabi** [comment]
@pettetmw
>They are saved as hd5.

If your output is a NumPy array then you can use either [`EvokedArray`](https://mne.tools/stable/generated/mne.EvokedArray.html?highlight=evoked%20array#mne.EvokedArray) to create an evoked or [`info`](https://mne.tools/stable/generated/mne.Info.html?highlight=info#mne.Info) along with `Epochs` class to clobber together an epochs object with the said array. The `hd5` format is a hangnail from when I was tooling around with `XDAWN`.

**Comment by pettetmw** [comment]
Sorry, I failed to specify: `.hd5` is the file type used by `mne.time_frequency.write_tfrs()` and `mne.time_frequency.read_tfrs()`, and also by the `save()` methods of the AverageTFR and EpochsTFR classes.  (I'm not sure what connection to `XDAWN` is?).
I'm not sure if that changes anything, but I just want to confirm: the scripts should coerce the data from `AverageTFR` object into `Evoked` object and save as `.evo` file, and vice-versa to read back again? Or should I stick with TFR save/read functions and `.hd5` files?
Also, do you want these in the `inverse_dir` with other evo files?

**Comment by ktavabi** [comment]
Forget what I said about `XDAWN`. Also, we should ditch `hd5`,  it's overkill for what we need: an `ipynb` at the end of the day that will run and print the figure in the manuscript so that any idiot with two thumbs and a keyboard can press `return` and get on their screen. To that end we need `Epoch` and `Evoked` class objects that once you have an array of values N-channels by N-time samples maybe reverse enginered and saved alongside the subject's `epochs` and `evoked` files on disk. Savvy?

**Comment by pettetmw** [comment]
Question regarding `write_subject_tfrs.py` and `viz_grp_assr.py`: Can I assume that `tfr_group.ipynb` notebook will reside and be executed in context where the following statements (or similar) lead to correct construction of mnefun filename?:

>include badbaby
>...
>% and later, e.g.,
>filename = op.join(defaults.datadir, subj, 'inverse','Oddball_%d-sss_eq_%s-ave.fif' % (lp, subj))

In particular, how should the path in `defaults.datadir` be made available? If not via `include badbaby`, should I instead use a string variable hard-coded to explicit `datadir` path?

**Comment by pettetmw** [comment]
(I enjoy these little conversations I have with myself :smile:)

**Comment by ktavabi** [comment]
I can tell you're just as fucked in the head as I am..😂

**Comment by ktavabi** [comment]
@pettetmw [Fig-SensorSpace.py](https://github.com/ktavabi/badbaby/commit/773467124c3183ab57c35fcf87640d0da778596c) in the Mismatch branch is what we need to aim for on the ASSR.

**Comment by pettetmw** [comment]
Do I need to `pip install -e badbaby` into my `mne` environment (or something similar) for `import badbaby` to work?

**Comment by ktavabi** [comment]
Yes

**Comment by pettetmw** [comment]
@ktavabi or @larsoner , I just noticed that my TFR calculation is reading `All_*-epo.fif`, which presumably contains "All" the epochs, not just the AM tone epochs that I want.  Can you point me to an example of how to pick the latter from the former, once I've read the file?

**Comment by larsoner** [comment]
`epochs['tone']` probably

https://mne.tools/dev/auto_tutorials/epochs/10_epochs_overview.html#subselecting-epochs

**Comment by pettetmw** [comment]
@ktavabi , `badbaby/processing/write_sbj_tfrs.py` approaches the "idiot with two thumbs" benchmark.  You'll need to be not-an-idiot enough to modify the value of `datadir` variable.  Individual `epochs` and `inverse` dirs will be populated with `-tfr.h5` files of TFR data. (h5 not optional if we want the frequency dimension). The cell at the end of the script will read all of these and create tidy csv.  Please LMK if it runs correctly for you.  If so, then we can revise if needed, and start adding code to visualize summaries?

**Comment by ktavabi** [comment]
@pettetmw I am cherry-picking gists from this PR into #19 and try to get it working on the database structure we've discussed in the past.

File: badbaby/processing/write_sbj_tfrs.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+# # %% loop over age and sbj
+mne.utils.set_config('MNE_USE_CUDA', 'true')
+for ig,g in enumerate(picks): # each group
+    for iss,ss in enumerate(g): # each subject id path
+
+        epo = mne.read_epochs( epops[ig][iss] )['tone']
+
+        #TODO: need to select 
+
```
Comment (ktavabi) [inline badbaby/processing/write_sbj_tfrs.py:57]: Begin-cherry-pick

File: badbaby/processing/write_sbj_tfrs.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        # Loop over and save tfr variables as ".h5", which allows us to include
+        # the frequency dimension (in addition to trials, channels, and time points)
+        for iv,vn in enumerate(tfr[0]): # tfr[0] is the list of tfr variable names
+            if vn[-3:] == 'evo':
+                tfr_fname = os.path.join( datadir, ss, ave_dir, 'All_'+lp+inv_tag+'_'+eq_tag+'_'+ss+vn+tfr_tag+'.h5' )
+            else:
+                tfr_fname = os.path.join( datadir, ss, epochs_dir, 'All_'+lp+inv_tag+'_'+ss+vn+tfr_tag+'.h5' )
+            #print(tfr_fname)
+            tfr[1][iv].save(tfr_fname,overwrite=True) # tfr[1] is the list of tfr object variables
+
```
Comment (ktavabi) [inline badbaby/processing/write_sbj_tfrs.py:111]: end cherry-pick

File: badbaby/processing/write_sbj_tfrs.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+ttab = [] # the table to save
+for ia,age in enumerate(age2gid.keys()):
+    ttab.append([])
+    for tfr in tfr_vnms:
+        # hereafter, "sid" is subject id; aslo make sure that "tfrp" has been created above
+        vn = '-'+tfr+'-evo'
+        tfr_fname_patt = os.path.join( datadir, '*'+grptags[ia], ave_dir, 'All_'+lp+inv_tag+'_'+eq_tag+'_'+'*'+grptags[ia]+vn+tfr_tag+'.h5' )
+        sidps = sorted(glob(tfr_fname_patt))
+        sidps = sidps[:4] # shorty version for quick testing
+        sids = [ os.path.basename(sidp)[-23:-16] for sidp in sidps ] # get id from full path
```
Comment (ktavabi) [inline badbaby/processing/write_sbj_tfrs.py:127]: This is ugly

File: badbaby/processing/write_sbj_tfrs.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        for iv,vn in enumerate(tfr[0]): # tfr[0] is the list of tfr variable names
+            if vn[-3:] == 'evo':
+                tfr_fname = os.path.join( datadir, ss, ave_dir, 'All_'+lp+inv_tag+'_'+eq_tag+'_'+ss+vn+tfr_tag+'.h5' )
+            else:
+                tfr_fname = os.path.join( datadir, ss, epochs_dir, 'All_'+lp+inv_tag+'_'+ss+vn+tfr_tag+'.h5' )
+            #print(tfr_fname)
+            tfr[1][iv].save(tfr_fname,overwrite=True) # tfr[1] is the list of tfr object variables
+
+# %% build CSV
+
```
Comment (ktavabi) [inline badbaby/processing/write_sbj_tfrs.py:113]: I am not quite sure what this cell does, but it's rather hard to read and has some exemplary bad variable naming.

File: badbaby/processing/write_sbj_tfrs.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+epochs_tag = params.epochs_tag
+inv_tag = params.inv_tag
+eq_tag = params.eq_tag
+#datadir = '/mnt/ilabs/tcirc/tone'
+#lp = '100'
+datadir = '/mnt/ilabs/badbaby/data'
+lp = '50'
+ave_dir = inverse_dir
+ave_tag = '-ave'
+tfr_tag = '-tfr'
```
Comment (ktavabi) [inline badbaby/processing/write_sbj_tfrs.py:28]: unnecessary

File: badbaby/processing/write_sbj_tfrs.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+        # compute the mean of the variances along real and imaginary axis
+        tVFFT = np.mean( np.stack( ( np.var( np.real(tfr_epo.data), 0 ), np.var( np.imag(tfr_epo.data), 0 ) ) ), 0 )
+        numerator = abs(tMFFT)
+        denominator = np.sqrt( tVFFT / ( tNTrl - 1 ) )
+        tcirc = (numerator / denominator)**2
+        tcirc_evo = itc_evo.copy()
+        tcirc_evo.data = tcirc
+        
+        tfr = [ [ '-tfr-epo', '-pow-epo', '-pow-evo', '-itc-evo', '-hot-evo', '-t2c-evo'],
```
Comment (ktavabi) [inline badbaby/processing/write_sbj_tfrs.py:93]: This is unorthodox. Python lists are mutable and not ideal as containers for ND data.

File: badbaby/processing/write_sbj_tfrs.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        
+        # power EpochsTFR
+        pow_epo = mne.time_frequency.tfr_morlet( epo, freqs=freqs, n_cycles=n_cycles, use_fft=True,
+                            return_itc=False, average=False, output='power')
+        
+        # power and ITC AverageTFR
+        pow_evo, itc_evo = mne.time_frequency.tfr_morlet( epo, freqs=freqs, n_cycles=n_cycles, use_fft=True,
+                            return_itc=True, average=True, output='power')
+        
+        tNTrl, tNCh, tNFrq, tNS = tfr_epo.data.shape
```
Comment (ktavabi) [inline badbaby/processing/write_sbj_tfrs.py:72]: DRY

File: badbaby/processing/write_sbj_tfrs.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+# # %% loop over age and sbj
+mne.utils.set_config('MNE_USE_CUDA', 'true')
+for ig,g in enumerate(picks): # each group
+    for iss,ss in enumerate(g): # each subject id path
+
+        epo = mne.read_epochs( epops[ig][iss] )['tone']
+
+        #TODO: need to select 
+
+        tNTrl, tNCh, tNS = epo.get_data().shape
```
Comment (ktavabi) [inline badbaby/processing/write_sbj_tfrs.py:58]: What the 🤬 does `t` stand for?

File: badbaby/processing/write_sbj_tfrs.py (hunk 4 above)
Comment (ktavabi) [inline badbaby/processing/write_sbj_tfrs.py:127]: 💩💩💩

**Review by ktavabi — COMMENTED** [review]
This definitely doesn't work on my end without editing. TBC

**Comment by ktavabi** [comment]
> @pettetmw I am cherry-picking gists from this PR into #19 and try to get it working on the database structure we've discussed in the past.

File: badbaby/processing/write_sbj_tfrs.py (hunk 4 above)
Comment (pettetmw) [reply badbaby/processing/write_sbj_tfrs.py:127]: why stop at three?

Comment (pettetmw) [reply badbaby/processing/write_sbj_tfrs.py:127]: @ktavabi, Just for my reference, is it the whole thing, or just line 127?  I really don't know how to name these files in an MNE-compliant way, so I could use some more specific guidance than poop emojis.

File: badbaby/processing/write_sbj_tfrs.py (hunk 8 above)
Comment (pettetmw) [reply badbaby/processing/write_sbj_tfrs.py:72]: I thought an implementation like yours would be more appropriate, but didn't get around to it... my bad.  I like the way you wrote it.

File: badbaby/processing/write_sbj_tfrs.py (hunk 7 above)
Comment (pettetmw) [reply badbaby/processing/write_sbj_tfrs.py:93]: good catch

File: badbaby/processing/write_sbj_tfrs.py (hunk 4 above)
Comment (ktavabi) [reply badbaby/processing/write_sbj_tfrs.py:127]: It's the whole thing, Mark, the setup is different from upstream granted, but I couldn't get it to work with a list of subjects. For naming anything in general, to make it compliant, you tack on `-tfr`. What you've done here is just asking for trouble in any circumstance. 
Also, it's taken way too long to get to here with the guidance I've provided so far, despite the constraint with emojis.

Comment (pettetmw) [reply badbaby/processing/write_sbj_tfrs.py:127]: Agreed.  For the record, I don't object to the emojis.  If the code doesn't work and isn't readable enough for you to debug, then three poops is being generous.

**Pull request closed without merging**